frontend/src/components/TenantDatabaseTable/DatabaseMatrix.vue

```diff
@@ -0,0 +1,254 @@
+    },
+    customClick: {
+      default: false,
+      type: Boolean,
+    },
+    name: {
+      type: String,
+      default: "",
+    },
+    showHeader: {
+      type: Boolean,
+      default: true,
+    },
+    databaseList: {
+      type: Object as PropType<Database[]>,
+      required: true,
+    },
+    environmentList: {
+      type: Array as PropType<Environment[]>,
+      required: true,
+    },
+    labelList: {
+      type: Array as PropType<Label[]>,
+      required: true,
+    },
+  },
+  emits: ["select-database"],
+  setup(props) {
+    // make "bb.environment" non-selectable because it was already specified to the x-axis
+    const selectableLabelList = computed(() => {
+      const excludes = new Set(["bb.environment"]);
+      return props.labelList.filter((label) => !excludes.has(label.key));
+    });
+
+    /**
+     * databases are grouped by `yAxisLabel` then by `xAxisLabel`
+     * for now, `xAxisLabel` will always be 'bb.environment'
+     */
+    const yAxisLabel = ref<LabelKeyType>();
+    const xAxisLabel = ref<LabelKeyType>("bb.environment");
+
+    const filteredXAxisValues = ref<LabelValueType[]>([]);
+    const filteredMatrices = ref<DatabaseMatrix[]>([]);
+
+    // find the default label key to firstBy (y-axis)
+    watchEffect(() => {
+      // "bb.environment" is excluded because it was specified to the x-axis
+      yAxisLabel.value = findDefaultGroupByLabel(
+        selectableLabelList.value,
+        props.databaseList
+      );
+    });
+
+    // pre-filtered y-axis values
+    const yAxisValues = computed((): string[] => {
+      const key = yAxisLabel.value;
+      if (!key) {
+        // y-axis is undefined
+        return [];
+      }
+
+      // order based on label.valueList
+      // plus one more "<empty value>"
+      const label = props.labelList.find((label) => label.key === key);
+      if (!label) return [];
+      return [...label.valueList, ""];
+    });
+
+    // pre-filtered x-axis values
+    const xAxisValues = computed(() => {
+      // order based on label.valueList
+      // plus one more "<empty value>"
+      const key = xAxisLabel.value;
+      const label = props.labelList.find((label) => label.key === key);
+      if (!label) return [];
+      return [...label.valueList, ""];
+    });
+
+    // first, group databases by `yAxisLabel` into some rows
+    const groupedRows = computed(() => {
+      const key = yAxisLabel.value;
+      if (!key) {
+        // y-axis is undefined
+        return [];
+      }
+      const dict = groupBy(props.databaseList, (db) => getLabelValue(db, key));
+      return yAxisValues.value.map((labelValue) => {
+        const databaseList = dict[labelValue] || [];
+        return {
+          labelValue,
+          databaseList,
+        };
+      });
+    });
+
+    // then, group each row by `xAxisLabel` into some columns
+    // now the `matrices` is pre-filtered (with empty rows or columns)
+    const matrices = computed((): DatabaseMatrix[] => {
```

Comment:
Please use xxxList to represent array. And this is also a classic example why I personally against using plural form, for matrix, there are actually 2 forms which both syntax correct.

Also simply suffix s is easier to make mistake because the only difference is an 's'.

Reply:
updated, PTAL